### PR TITLE
P2314R4 Character sets and encodings

### LINKED ISSUE #4971: [2021-10 CWG Motion 10] P2314R4 Character sets and encodings
https://wiki.edg.com/pub/Wg21virtual2021-10/StrawPolls/p2314r4.html

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #4971
Fixes cplusplus/papers#998

### CODE REVIEW

**Comment by hubert-reinterpretcast** [comment]
@jensmaurer @jwakely, this is missing an editorial update of the cross-reference from [fs.path.type.cvt] for the term "execution wide-character set". The cross-reference should point to [character.seq] and not [lex.charset].

**Comment by jensmaurer** [comment]
@hubert-reinterpretcast , fixed.

File: source/lib-intro.tex [hunk 1]
```diff
@@ -646,14 +646,29 @@
 
 \begin{itemize}
 \item
+Properties specified as locale-specific may change during program execution
+by a call to \tcode{setlocale(int, const char*)}\iref{clocale.syn}, or
+by a change to a \tcode{locale} object,
+as described in \ref{locales} and \ref{input.output}.
+\item
+The \defnadj{execution}{character set} and
+the \defnadj{execution}{wide-character set}
```
Comment (burblebee) [inline source/lib-intro.tex:655]: Problem with wording in paper: I can't tell what these terms mean based on this definition. What are they? Should we open an issue?

File: source/compatibility.tex [hunk 2]
```diff
@@ -786,8 +786,8 @@
 semantics in this revision of \Cpp{}. Implementations may choose to
 translate trigraphs as specified in \CppXIV{} if they appear outside of a raw
 string literal, as part of the \impldef{mapping from physical source file characters
-to basic source character set} mapping from physical source file characters to
-the basic source character set.
+to translation character set} mapping from physical source file characters to
+the translation character set.
```
Comment (burblebee) [inline source/compatibility.tex:790]: The paper asked to keep "basic" here; should this change be in a separate commit?

**Review by burblebee — CHANGES_REQUESTED** [review]

File: source/lex.tex [hunk 3]
```diff
@@ -342,20 +401,18 @@
...
-matches the last category, the behavior is undefined. Preprocessing tokens can be
-separated by
+Preprocessing tokens can be separated by
 \indextext{whitespace}%
 whitespace;
 \indextext{comment}%
 this consists of comments\iref{lex.comment}, or whitespace
-characters (space, horizontal tab, new-line, vertical tab, and
-form-feed), or both. As described in \ref{cpp}, in certain
+characters (U+0020 SPACE, U+0009 CHARACTER TABULATION, new-line, U+000B LINE TABULATION, and U+000C FORM FEED), or both. As described in \ref{cpp}, in certain
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:415]: P2348 Whitespaces Wording Revamp cplusplus/papers#1027 will address this.

File: source/lex.tex [hunk 4]
```diff
@@ -224,25 +215,74 @@
...
+\end{note}
+
+\begin{floattable}{Basic character set}{lex.charset.basic}{lll}
+\topline
+\lhdrx{2}{character} & \rhdr{glyph} \\ \capsep
+U+0009 & CHARACTER TABULATION & \\
+U+000B & LINE TABULATION & \\
+U+000C & FORM FEED (FF) & \\
+U+0020 & SPACE & \\
+U+000A & LINE FEED (LF) & new-line \\
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:250]: P2348 Whitespaces Wording Revamp cplusplus/papers#1027 will address this.

File: source/lex.tex [hunk 5]
```diff
@@ -642,14 +699,14 @@
...
-    universal-character-name \textnormal{of class XID_Start}
+    \textnormal{an element of the translation character set of class XID_Start}
 \end{bnf}
 
 \begin{bnf}
 \nontermdef{identifier-continue}\br
     digit\br
     nondigit\br
-    universal-character-name \textnormal{of class XID_Continue}
+    \textnormal{an element of the translation character set of class XID_Continue}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:709]: See the text that follows; those are from UAX something.

File: source/lex.tex [hunk 6]
```diff
@@ -1300,7 +1357,7 @@
 The examples in \tref{lex.ccon.literal}
 for non-encodable ordinary and wide character literals assume that
 the specified character lacks representation in
-the execution character set or execution wide-character set, respectively, or
+the ordinary literal encoding or wide literal encoding, respectively, or
 that encoding it would require more than one code unit.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:1361]: "it" = "character"

[the remark this answers is not in the document]
Comment (hubert-reinterpretcast) [reply source/lex.tex:1361]: > "it" = "character"

... more specifically, "the specified character" previously mentioned.

File: source/basic.tex [hunk 7]
```diff
@@ -4911,10 +4910,10 @@
 
 \pnum
 \indextext{type!integral}%
-Types
-\keyword{bool},
-\keyword{char}, \keyword{wchar_t},
-\keyword{char8_t}, \keyword{char16_t}, \keyword{char32_t}, and
+The types \keyword{char}, \keyword{wchar_t},
+\keyword{char8_t}, \keyword{char16_t}, and \keyword{char32_t}
+are collectively called character types.
```
Comment (jensmaurer) [inline source/basic.tex:4915]:
```suggestion
are collectively called \defnadjx{character}{types}{type}.
```

File: source/lex.tex [hunk 8]
```diff
@@ -224,25 +215,74 @@
...
+\item
+each character named by ISO/IEC 10646,
+as identified by its unique UCS scalar value, and
+\item
+a distinct character for each UCS scalar value
+where no named character is assigned.
+\end{itemize}
+\begin{note}
+ISO/IEC 10646 code points are integers
+in the range  $[0, \mathrm{10FFFF}]$ (hexadecimal).
```
Comment (jensmaurer) [inline source/lex.tex:229]:
```suggestion
in the range $[0, \mathrm{10FFFF}]$ (hexadecimal).
```

File: source/lex.tex [hunk 9]
```diff
@@ -224,25 +215,74 @@
...
+A surrogate code point is a value
+in the range $[\mathrm{D800}, \mathrm{DFFF}]$ (hexadecimal).
+A UCS scalar value is any code point that is not a surrogate code point.
+\end{note}
+
+\pnum
+The \defnadj{basic}{character set} is a subset of the translation character set,
+consisting of 96 characters as specified in \tref{lex.charset.basic}.
+\begin{note}
+Unicode short names are given only as a means to identifying the character;
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:239]: The penultimate link shows "That's why I wanted to have it, to show it -- not use it as a means to painting, but use painting as a means to photography". "
which is exactly the phrasing I'm using here. "I'm looking forward to seeing you" is similar.

File: source/lex.tex [hunk 10]
```diff
@@ -260,54 +300,74 @@
...
+encodes each element of the basic literal character set as
+a single code unit with non-negative value,
+distinct from the code unit for any other such element.
+\begin{note}
+A character not in the basic literal character set
+can be encoded with more than one code unit;
+the value of such a code unit can be the same as
+that of a code unit for an element of the basic literal character set.
+\end{note}
+The U+0000 NULL character is encoded as the value 0.
```
Comment (jensmaurer) [inline source/lex.tex:360]:
```suggestion
The U+0000 NULL character is encoded as the value \tcode{0}.
```

File: source/lex.tex [hunk 11]
```diff
@@ -260,54 +300,74 @@
...
+No other element of the translation character set
+is encoded with a code unit of value 0.
```
Comment (jensmaurer) [inline source/lex.tex:362]:
```suggestion
is encoded with a code unit of value \tcode{0}.
```

File: source/lex.tex [hunk 12]
```diff
@@ -260,54 +300,74 @@
...
+The code unit value of each decimal digit character after the digit 0 (U+0030)
```
Comment (jensmaurer) [inline source/lex.tex:363]:
```suggestion
The code unit value of each decimal digit character after the digit \tcode{0} (U+0030)
```

File: source/lex.tex [hunk 13]
```diff
@@ -642,14 +699,14 @@
 \begin{bnf}
 \nontermdef{identifier-start}\br
     nondigit\br
-    universal-character-name \textnormal{of class XID_Start}
+    \textnormal{an element of the translation character set of class XID_Start}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:702]: See the text that follows; those are from UAX something.

File: source/lex.tex [hunk 14]
```diff
@@ -1614,9 +1672,10 @@
 
 \begin{bnf}
 \nontermdef{d-char}\br
-    \textnormal{any member of the basic source character set except:}\br
-    \bnfindent\textnormal{space, the left parenthesis \terminal{(}, the right parenthesis \terminal{)}, the backslash \terminal{\textbackslash}, and the control characters}\br
-    \bnfindent\textnormal{representing horizontal tab, vertical tab, form feed, and newline.}
+    \textnormal{any member of the basic character set except:}\br
+    \bnfindent\textnormal{U+0020 SPACE, U+0028 LEFT PARENTHESIS, U+0029 RIGHT PARENTHESIS, U+005C REVERSE SOLIDUS,
+    U+0009 CHARACTER TABULATION, U+000B LINE TABULATION, U+000C FORM FEED (FF),}\br
+    \bnfindent\textnormal{and newline.}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:1678]: This is pre-existing. If you feel this should be addressed, please open a separate issue with reference to P2348 Whitespaces Wording Revamp cplusplus/papers#1027

File: source/lex.tex [hunk 15]
```diff
@@ -1729,19 +1788,45 @@
...
+Any other combinations are ill-formed.
 \begin{note}
-This concatenation is an interpretation, not a conversion.
-Because the interpretation happens in translation phase 6
-(after the string literal contents have been encoded in
-the \grammarterm{string-literal}'s associated character encoding),
-a \grammarterm{string-literal}'s initial rawness
-has no effect on the interpretation or well-formedness of the concatenation.
+A \grammarterm{string-literal}'s rawness has
+no effect on the determination of the common \grammarterm{encoding-prefix}.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:1803]: The author (me) has confirmed that's what we want to say. (This is a note, so fixing it is fairly easy, too.)(

File: source/lex.tex [hunk 16]
```diff
@@ -1729,19 +1788,45 @@
...
 \end{note}
+
+\pnum
+In translation phase 6\iref{lex.phases},
+adjacent \grammarterm{string-literal}s are concatenated.
+The lexical structure and grouping of
+the contents of the individual \grammarterm{string-literal}s is retained.
+\begin{example}
+\begin{codeblock}
+"\xA" "B"
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:1813]: I don't know.

File: source/lex.tex [hunk 17]
```diff
@@ -1729,19 +1788,45 @@
...
+the code unit \tcode{'\textbackslash xA'} and the character \tcode{'B'}
+after concatenation
+(and not the single code unit \tcode{'\textbackslash xAB'}).
+Similarly,
+\begin{codeblock}
+"(\u00)" "41"
+\end{codeblock}
+represents six characters,
+starting with a backslash and ending with the digit \tcode{1}
+(and not the single character \tcode{"A"}
```
Comment (jensmaurer) [inline source/lex.tex:1825]:
```suggestion
(and not the single character \tcode{'A'}
```

File: source/lex.tex [hunk 18]
```diff
@@ -1799,8 +1873,9 @@
 String literal objects are initialized with
 the sequence of code unit values
 corresponding to the \grammarterm{string-literal}'s sequence of
-\grammarterm{s-char}s (for a non-raw string literal) and
-\grammarterm{r-char}s (for a raw string literal)
+\grammarterm{s-char}s (originally from non-raw- string literals) and
```
Comment (jensmaurer) [inline source/lex.tex:1876]:
```suggestion
\grammarterm{s-char}s (originally from non-raw string literals) and
```

File: source/lib-intro.tex (hunk 1 above)
Comment (jensmaurer) [reply source/lib-intro.tex:655]: Feel free to file an LWG issue, but this just moves these not-quite definitions from the core wording to the library wording, so I'd say this is pre-existing.

File: source/lex.tex (hunk 10 above)
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:360]: Yeah, we need to come with some consistent formatting for those \U+nnnn things.

File: source/lex.tex [hunk 19]
```diff
@@ -1812,17 +1887,13 @@
 is encoded to a code unit sequence
 using the \grammarterm{string-literal}'s associated character encoding.
 If a character lacks representation in the associated character encoding, then:
```
Comment (jensmaurer) [inline source/lex.tex:1889]:
```suggestion
If a character lacks representation in the associated character encoding,
```

File: source/lex.tex [hunk 20]
```diff
@@ -1602,8 +1660,8 @@
 
 \begin{bnf}
 \nontermdef{r-char}\br
-    \textnormal{any member of the source character set, except a right parenthesis \terminal{)} followed by}\br
-    \bnfindent\textnormal{the initial \grammarterm{d-char-sequence} (which may be empty) followed by a double quote \terminal{"}.}
+    \textnormal{any member of the translation character set, except a U+0029 RIGHT PARENTHESIS followed by}\br
+    \bnfindent\textnormal{the initial \grammarterm{d-char-sequence} (which may be empty) followed by a U+0022 QUOTATION MARK.}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:1664]: Without the indentation, it can be confused with another option for the grammar non-terminal.

File: source/lex.tex [hunk 21]
```diff
@@ -1614,9 +1672,10 @@
 
 \begin{bnf}
 \nontermdef{d-char}\br
-    \textnormal{any member of the basic source character set except:}\br
-    \bnfindent\textnormal{space, the left parenthesis \terminal{(}, the right parenthesis \terminal{)}, the backslash \terminal{\textbackslash}, and the control characters}\br
-    \bnfindent\textnormal{representing horizontal tab, vertical tab, form feed, and newline.}
+    \textnormal{any member of the basic character set except:}\br
+    \bnfindent\textnormal{U+0020 SPACE, U+0028 LEFT PARENTHESIS, U+0029 RIGHT PARENTHESIS, U+005C REVERSE SOLIDUS,
+    U+0009 CHARACTER TABULATION, U+000B LINE TABULATION, U+000C FORM FEED (FF),}\br
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex]: Fixed

File: source/lex.tex [hunk 22]
```diff
@@ -342,20 +401,18 @@
 (\grammarterm{import-keyword}, \grammarterm{module-keyword}, and \grammarterm{export-keyword}),
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lex.tex:405]: Oops, thanks for noticing.

File: source/lex.tex (hunk 10 above)
[the remark this answers is not in the document]
Comment (burblebee) [reply source/lex.tex]: Opened #5032.

File: source/lex.tex (hunk 20 above)
[the remark this answers is not in the document]
Comment (burblebee) [reply source/lex.tex]: Yes, I agree.  But it looks wrong visually.

File: source/lex.tex [hunk 23]
```diff
@@ -224,25 +215,74 @@
...
+The \defnadj{basic}{character set} is a subset of the translation character set,
+consisting of 96 characters as specified in \tref{lex.charset.basic}.
+\begin{note}
+Unicode short names are given only as a means to identifying the character;
+the numerical value has no other meaning in this context.
+\end{note}
+
+\begin{floattable}{Basic character set}{lex.charset.basic}{lll}
+\topline
+\lhdrx{2}{character} & \rhdr{glyph} \\ \capsep
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/lex.tex:245]: Yes; this just seemed like a case where we might want to break from that convention.

File: source/lib-intro.tex (hunk 1 above)
Comment (timsong-cpp) [reply source/lib-intro.tex:655]: The "implementation-defined" part was lost in the move. Is that intended?

Comment (jensmaurer) [reply source/lib-intro.tex:655]: The paper shows the "implementation-defined" part as removed. I don't think I intended to remove "implementation-defined", although it serves little purpose given that the only observable thing here is the encoding (which is locale-specific).

File: source/lib-intro.tex [hunk 24]
```diff
@@ -646,14 +646,29 @@
 
 \begin{itemize}
 \item
+Properties specified as locale-specific may change during program execution
```
Comment (tkoeppe) [inline source/lib-intro.tex:649]: @jensmaurer: the paper italicizes "locale-specific" here; do we want to keep that?

Comment (jensmaurer) [reply source/lib-intro.tex]: Yes, fixed.

File: source/lib-intro.tex (hunk 1 above)
Comment (jensmaurer) [reply source/lib-intro.tex:656]: @tkoeppe, what do you think?

File: source/lex.tex [hunk 25]
```diff
@@ -68,22 +68,11 @@
 \indextext{character!source file}%
 \indextext{character set!basic source}%
 Physical source file characters are mapped, in an
-\impldef{mapping physical source file characters to basic source character set} manner,
-to the basic source character set (introducing new-line characters for end-of-line
-indicators) if necessary.
+\impldef{mapping physical source file characters to translation character set} manner,
+to the translation character set (introducing new-line characters for end-of-line
```
Comment (tkoeppe) [inline source/lex.tex:72]: @jensmaurer: Not in the paper, but do you think a reference would be useful here, since this is the first time we see the term "translation character set"?

Comment (jensmaurer) [reply source/lex.tex]: Fixed. Please squash "fixup" commits when merging.

File: source/lib-intro.tex (hunk 1 above)
Comment (timsong-cpp) [reply source/lib-intro.tex:656]: Not particularly, but I think it helps with the "what these character sets are" question. We can sort this out in the LWG issue Dawn filed.

File: source/compatibility.tex (hunk 2 above)
Comment (JohelEGP) [reply source/compatibility.tex:805]: It's not really fixed.

File: source/lex.tex [hunk 26]
```diff
@@ -68,22 +68,11 @@
 \indextext{character!source file}%
 \indextext{character set!basic source}%
 Physical source file characters are mapped, in an
-\impldef{mapping physical source file characters to basic source character set} manner,
-to the basic source character set (introducing new-line characters for end-of-line
-indicators) if necessary.
+\impldef{mapping physical source file characters to translation character set} manner,
```
Comment (hubert-reinterpretcast) [inline source/lex.tex:71]: @jensmaurer, this is subtly different from the other cross-reference for the same concept (the other index entry has "mapping from" instead of just "mapping").
